kyo-core/shared/src/main/scala/kyo/Queue.scala

```diff
@@ -339,42 +353,84 @@ object Queue:
         def peek()(using AllowUnsafe): Result[Closed, Maybe[A]]
         def drain()(using AllowUnsafe): Result[Closed, Chunk[A]]
         def close()(using Frame, AllowUnsafe): Maybe[Seq[A]]
+        def closeAwaitEmpty()(using Frame, AllowUnsafe): Fiber.Unsafe[Nothing, Boolean]
         def closed()(using AllowUnsafe): Boolean
         final def safe: Queue[A] = this
     end Unsafe
 
     /** WARNING: Low-level API meant for integrations, libraries, and performance-sensitive code. See AllowUnsafe for more details. */
     object Unsafe:
 
+        private enum State derives CanEqual:
+            case Open
+            case HalfClosed(p: Promise.Unsafe[Nothing, Boolean], r: Result.Error[Closed])
+            case FullyClosed(r: Result.Error[Closed])
+        end State
+
         sealed abstract private class Closeable[A](initFrame: Frame) extends Unsafe[A]:
             import AllowUnsafe.embrace.danger
-            final protected val _closed = AtomicRef.Unsafe.init(Maybe.empty[Result.Error[Closed]])
+            final private val state = AtomicRef.Unsafe.init[State](State.Open)
 
             final def close()(using frame: Frame, allow: AllowUnsafe) =
                 val fail = Result.Failure(Closed("Queue", initFrame))
-                Maybe.when(_closed.compareAndSet(Maybe.empty, Maybe(fail)))(_drain())
+                Maybe.when(state.compareAndSet(State.Open, State.FullyClosed(fail)))(_drain())
             end close
 
-            final def closed()(using AllowUnsafe) = _closed.get().isDefined
+            final def closeAwaitEmpty()(using frame: Frame, allow: AllowUnsafe): Fiber.Unsafe[Nothing, Boolean] =
+                val fail = Result.Failure(Closed("Queue", initFrame))
+                val p    = Promise.Unsafe.init[Nothing, Boolean]()
+                if state.compareAndSet(State.Open, State.HalfClosed(p, fail)) then
+                    handleHalfOpen()
+                    p
+                else
+                    Fiber.Unsafe.init(Result.succeed(false))
+                end if
+            end closeAwaitEmpty
+
+            final def closed()(using AllowUnsafe) =
+                state.get().isInstanceOf[State.FullyClosed]
 
             final def drainUpTo(max: Int)(using AllowUnsafe): Result[Closed, Chunk[A]] = op(_drain(Maybe.Present(max)))
 
             final def drain()(using AllowUnsafe): Result[Closed, Chunk[A]] = op(_drain())
 
             protected def _drain(max: Maybe[Int] = Maybe.Absent): Chunk[A]
+            protected def _isEmpty(): Boolean
+
+            private def opClosed: Maybe[Result.Error[Closed]] =
+                state.get() match
+                    case State.FullyClosed(r) => Present(r)
+                    case _                    => Absent
 
             protected inline def op[A](inline f: => A): Result[Closed, A] =
```

Comment:
The half closed state allows all operations (`op` method`) and fails in case of offers or adds (`offerOp`)